docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses very small memory. 
+
+For inactive IndexReaders, if the inactive duration (`now.Sub(lastQueryTime)`) greats than a configured duration (e.g. 15 mins), then IndexReader will remove in-memory objects and close file descriptor and mmap in IndexFile.  This will reduce memory usage.
+
+**NOTE**: The `pages` bitmap in BucketFile should always be kept in memory until the server is closed.
+
+### Preload blocks on startup
+
+The latest blocks will be queried mostly. 
+
+Preload `the last N hours` (configurable) blocks on startup.
+This will fetch index files from object storage and decode bytes into in-memory objects.
+
+### Local Files
+
+Because the max size of a index file is 64 GiB, and  the max size of a chunks file is 128 MiB.
+
+So we can set the size of a local index file to 64 GiB, set  the size of a local chunks file to 128 MiB
+
+The size is used to create a sparse file and mmap a file.
+
+### Object storage byte ranges
+
+Currently, different provider clients implement byte ranges  differently. We need to wrap their implementations and export the same api.
+
+ I think byte ranges in [minio-go](https://github.com/minio/minio-go/blob/master/api-get-options.go#L99-L128) is a good example:
+
+1. Fetch last N bytes:     
+     `Range{start:0, end: -N}` ->  Request header`Range: bytes=-N`     
+     **NOTE**: will not write to local file because we do not know the offset.
+2. Fetch everything starting from `start`:    
+    `Range{start:N, end: 0}	`  ->  Request heade `Range: bytes=N-` 
+
+3. Fetch everything starting at `start` till  `end`:    
+   `Range{start:N, end: M}` -> Request header `Range: bytes=M-N` 
+
+if providers support multiple ranges, likes `Range: bytes=500-700,60000-61000`, we need to implement a `BucketReader.GetRanges` API, then we can query multiple ranges in one request.
+
+See https://tools.ietf.org/html/rfc7233#section-2.1 for reference.
+
+## Risks
+
+Ensure all providers support `fetch last N bytes` and `fetch everything starting from offset`. Otherwise, we need to know the size of remote index file before fetching TOC block. We may get the size by fetching metadata from object storage.
+
+### Work plan
+
+- A) Investigate and Implement `BucketReader.GetRange`  and  `BucketReader.GetRanges`
+  - we have implemented  `BucketReader.GetRange` , check if it has the ability to fetch last N bytes
+  - implement `BucketReader.GetRanges` if providers support
+- B) Implement `BucketFile`
+  
+- C) Use `BucketFile` for `BucketIndexReader` and `BucketchunkReader`
+
+- D) Add Flags to enable `BucketFile` 
+
+### Future work
+
+Design a new index file for [postings offset entries](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset entries](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) for sequential and  random accesses.
```

Comment:
Future work means not `next steps to implement this designed idea`  - it's rather: What's next if we implement this design?

Reply:
This needs to design a new search algorithm and file disk format. 

Although its goal is to reduce memory, it is independent and can be described in the next design.

Should describe details in this design？

Reply:
Sorry I thought you mean something else. 

No, it should not be part of the design. I would state it as `Consider changing index file design for (...)` as we first need to agree on this. I am quite sure we would like to be compliant with native TSDB (: